deluge/httpdownloader.py

```diff
@@ -31,13 +33,38 @@
 log = logging.getLogger(__name__)
 
 
-class HTTPDownloader(client.HTTPDownloader):
+class BodyHandler(HTTPClientParser):
+    def __init__(self, request, finished, length, agent):
+        HTTPClientParser.__init__(self, request, finished)
+        self.agent = agent
+        self.finished = finished
+        self.total_length = length
+        self.current_length = 0
+        self.data = b''
+
+    def dataReceived(self, data):  # NOQA: N802
+        self.current_length += len(data)
+        if self.agent.decoder:
+            data = self.agent.decoder.decompress(data)
+        self.data += data
+        if self.agent.part_callback:
+            self.agent.part_callback(data, self.current_length, self.total_length)
+
+    def connectionLost(self, reason):  # NOQA: N802
+        with open(self.agent.filename, 'wb') as _file:
+            _file.write(self.data)
+        self.finished.callback(self.agent.filename)
+        self.state = u'DONE'
+        HTTPClientParser.connectionLost(self, reason)
+
+
+class HTTPDownloaderAgent(client.Agent):
```

Comment:
I am not sure about subclassing Agent, I would prefer to pass agent as an arg so follow the style of the other web client agents: https://github.com/twisted/twisted/blob/trunk/src/twisted/web/client.py#L1941

This would mean that url and headers should be passed to request instead of HTTPDownloaderAgent